### ISSUE TITLE
How to activate nogo

### ISSUE THREAD

**Issue by Globegitter** [bazelbuild/rules_go#1777]
The documentation mentions the release of `nogo` which sounds very interesting but does not directly state on how to activate it. Does it just run by default? Or does it need to be put into a top level `BUILD.bazel` file similar to `gazelle`? Have not been able to find any mention of that in the docs.

**Comment by justinsb** [comment]
I'd echo this - I tried to just get go vet working and couldn't get it to work.  It was complaining about a missing `/tmp/govet.cfg` file (IIRC) which I thought was a good sign, but then later it "just stopped" complaining but also wasn't picking up a go vet problem (of the `-printf` class), so I assume it stopped complaining because it stopped running.

My WIP branch is here if it's relevant: https://github.com/justinsb/etcd-manager/commit/169cbfad24b34a4095139247b214da6721f10293

**Comment by jayconrod** [comment]
@Globegitter [go/nogo.rst](https://github.com/bazelbuild/rules_go/blob/master/go/nogo.rst) is the best place to start. Basically, you need to declare a `nogo` target somewhere, then point to it from `go_register_toolchains`. It will run automatically for each Go package that gets built.

@stjj89 Do you have bandwidth today to look into @justinsb's example?

**Comment by justinsb** [comment]
Ah - thanks @jayconrod  - the "point to it from go_register_toolchains" makes it obvious to me what I'm doing when I declare it in go_register_toolchains, and explains why I only have to put it in the root.  (That would be a great phrasing for the docs, IMO)

I have a theory that I'm running in parallel and all the tasks are writing to the same `/tmp/vet.cfg` file (and deleting it).  i.e. we should use a genuine `ioutil.TempFile` or a vet.cfg file in a tempdir if the name matters.  I think this could explain sometimes seeing file-not-found, and maybe I'm not getting vet errors where I think I should be because vet.cfg keeps being rewritten with different packages.  That said, this machine is only 2 cores, so it's not a great theory...

But `bazel build --jobs=1 //pkg/...` does indeed complete without the file-not-found errors, though I still have yet to have a `go vet` failure...

**Comment by justinsb** [comment]
I added an egregious `if true || true` (borrowing the idea from the tests for nogo) and it does correctly no-go me :-)  https://github.com/kopeio/etcd-manager/pull/134/files  

And I'm pretty sure the single file is an issue, because with jobs=1 I didn't see problems, but without jobs=1 I saw: `vet: open /tmp/vet.cfg: no such file or directory`

and the real smoking gun: `vet: parsing vet config /tmp/vet.cfg: unexpected end of JSON input`

I'll try to figure out why the `-printf` warning isn't firing for me, but I think that's a separate issue.

This is awesome BTW - thank you so much for adding this :-)

**Comment by stjj89** [comment]
@justinsb Thanks for reporting this issue. @jayconrod it sounds like you're already working on a fix, so I'll leave that in your capable hands.

@Globegitter the [nogo](https://github.com/bazelbuild/rules_go/blob/master/go/nogo.rst) docs are currently linked in the table of contents of the [main README](https://github.com/bazelbuild/rules_go/blob/master/README.rst) as "nogo build-time code analysis". I can see how it's a little hard to find; perhaps we can do better. May I know how and where you tried looking for documentation, and where you would expect to find information about this tool?

**Comment by Globegitter** [comment]
@stjj89 I saw the documentation, but I expected something like a "setup" where it would explain me exactly how to set things up, ideally with some copy&paste snippets. But the real missing key was that I could not find any reference of having to add this to `go_register_toolchains`. Am I just missing that part?

**Comment by stjj89** [comment]
Right now this instruction is nestled at the end of the [Writing and registering analyzers](https://github.com/bazelbuild/rules_go/blob/master/go/nogo.rst#writing-and-registering-analyzers) section, followed by a code snippet:

> Finally, the `nogo`_ target must be passed to ``go_register_toolchains``
in your ``WORKSPACE`` file.

I can see how you might have missed that, especially if you don't intend to write your own analyzers and skipped directly to the section on running vet.

Thanks for the feedback. I'll send out a PR to restructure this doc so that the first section tells you very directly what you need to do to run `nogo`.

**Comment by Globegitter** [comment]
Both PRs have been merged now and I activated it already in our project (only vet for now, but will start to try some of the other ones as well)
